### PR TITLE
Prevent bookings with same start / stop time

### LINKED ISSUE #2342: Prevent bookings with same start / stop time
**Is your feature request related to a problem? Please describe.**
Bookings with same start and stop time are making no sense, but are now possible

**Describe the solution you'd like**
Prevent these bookings.
If there is another sense behind this behavior make it optional via settings.

### PR DESCRIPTION
**Pull request by Wolfimschafspelz**
## Description
- Makes it impossible to create bookings with the same start / stop time when unchecking a box in the system settings

grafik

grafik

- closes #2342

## Types of changes
- [x] New feature (non-breaking change which adds functionality

## Checklist
- [x] I verified that my code applies to the guidelines (`composer code-check`)
- [x] I agree that this code is used in Kimai and will be published under the [MIT license](https://github.com/kevinpapst/kimai2/blob/master/LICENSE)

### CODE REVIEW

File: src/Validator/Constraints/TimesheetZeroLengthValidator.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        if (!\is_object($timesheet) || !($timesheet instanceof TimesheetEntity)) {
+            throw new UnexpectedTypeException($timesheet, TimesheetEntity::class);
+        }
+
+        if ($this->configuration->isTimesheetAllowZeroLength()) {
+            return;
+        }
+
+        if ($timesheet->getDuration() == 0) {
+            $this->context->buildViolation('The duration cannot be zero.')
```
Comment (kevinpapst) [inline src/Validator/Constraints/TimesheetZeroLengthValidator.php:49]:
```suggestion
            $this->context->buildViolation($constraint->message)
```

File: src/Validator/Constraints/TimesheetZeroLength.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+namespace App\Validator\Constraints;
+
+final class TimesheetZeroLength extends TimesheetConstraint
+{
+    public const ZERO_LENGTH_ERROR = 'kimai-timesheet-zero-length-01';
+
+    protected static $errorNames = [
+        self::ZERO_LENGTH_ERROR => 'The Duration can not be zero.',
```
Comment (kevinpapst) [inline src/Validator/Constraints/TimesheetZeroLength.php:17]:
```suggestion
        self::ZERO_LENGTH_ERROR => 'Duration cannot be zero.',
```

File: src/Validator/Constraints/TimesheetZeroLength.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    ];
+
+    public $message = 'The Duration can not be zero.';
```
Comment (kevinpapst) [inline src/Validator/Constraints/TimesheetZeroLength.php:20]:
```suggestion
    public $message = 'Duration cannot be zero.';
```

**Review by kevinpapst — CHANGES_REQUESTED** [review]
Thank @Wolfimschafspelz 👍 

We need to add some missing translations, so the show up in Weblate for the translation to all other languages.

In `validators.en.xlf` the key `Duration cannot be zero.` with the same translations.

And your username suggests that you can add the German translations for the two new messages 😄 ?!?

**Review by kevinpapst — CHANGES_REQUESTED** [review]
Hey, after giving another thought, I'd like to see two more changes:
- replace "Length" with "Duration" in all places
- add a unit test for the validator

Edit: the last fixes were exactly what I had in mind, thank you 👍

File: tests/Validator/Constraints/TimesheetZeroDurationValidatorTest.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        $timesheet = new Timesheet();
+        $timesheet->setBegin($begin);
+        $timesheet->setEnd($begin);
+
+        return $timesheet;
+    }
+
+    public function testZeroDurationIsDisallowed()
+    {
+        $timesheet = prepareTimesheet();
```
Comment (kevinpapst) [inline tests/Validator/Constraints/TimesheetZeroDurationValidatorTest.php:72]:
```suggestion
        $timesheet = $this->prepareTimesheet();
```

File: tests/Validator/Constraints/TimesheetZeroDurationValidatorTest.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+        $this->buildViolation('Duration cannot be zero.')
+            ->atPath('property.path.duration')
+            ->setCode(TimesheetZeroDuration::ZERO_DURATION_ERROR)
+            ->assertRaised();
+    }
+
+    public function testZeroDurationIsAllowed()
+    {
+        $timesheet = prepareTimesheet();
```
Comment (kevinpapst) [inline tests/Validator/Constraints/TimesheetZeroDurationValidatorTest.php:84]:
```suggestion
        $this->validator = $this->createMyValidator(true);
        $this->validator->initialize($this->context);

        $timesheet = $this->prepareTimesheet();
```

File: tests/Validator/Constraints/TimesheetZeroDurationValidatorTest.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        $this->validator->validate($timesheet, new TimesheetZeroDuration(['message' => 'myMessage']));
+
+        self::assertEmpty($this->context->getViolations());
```
Comment (kevinpapst) [inline tests/Validator/Constraints/TimesheetZeroDurationValidatorTest.php:88]:
```suggestion
        $this->assertNoViolation();
```

**Review by kevinpapst — COMMENTED** [review]
There are some other issues in existing tests, due to the added configuration. I can solve them later on, if you don't find the solution.

**Comment by kevinpapst** [comment]
Please run `composer codestyle-fix` and `composer phpstan` before committing.

**Comment by Wolfimschafspelz** [comment]
Hey,
There are one error and one failed test case in the integration tests remaining which I was unable to resolve.

**Review by kevinpapst — APPROVED** [review]
Looks good to me 👍  

Thank you @Wolfimschafspelz for being patient 😄

**Pull request merged**